PolyEngine/Core/Src/RTTI/RTTI2.hpp

```diff
@@ -0,0 +1,208 @@
+#pragma once
+#include <any>
+#include <typeindex>
+#include <vector>
+#include <memory>
+#include <unordered_map>
+#include <algorithm>
+#include <stdexcept>
+
+namespace Poly
+{
+    constexpr auto List = [](auto ...xs) {
+        return [=](auto access) { return access(xs...); };
+    };
+
+    template <typename F, typename X, typename ... XS>
+    void _runList(F f, X x, XS... xs)
+    {
+        f(x);
+        _runList(f, xs...);
+    }
+
+    template <typename F, typename X>
+        void _runList(F f, X x) { f(x); }
+
+    constexpr auto runList = [](auto list, auto f) {
+        list([f](auto... xs){ _runList(f, xs...); });
+    };
+    namespace RTTI2
+    {
+        template <typename T> struct RTTIinfo {
+            constexpr static auto info = List();
+        };
+
+        using AttrMap = std::unordered_map<std::type_index, std::any>;
+
+        template <typename T> struct SimpleAttribute
+        {
+            using container = std::vector<T>;
+            static void add(AttrMap& m, T x)
+            {
+                auto i = m.find(typeid(T));
+                if(i == m.end())
+                    i = m.emplace(typeid(T), container{}).first;
+                std::any_cast<container>(&i->second)->emplace_back(x);
+            }
+        };
+
+        template <typename T> struct DerivedAttribute
+        {
+            using container = std::vector<std::shared_ptr<T>>;
+            template <typename TT> static void add(AttrMap& m, TT x)
+            {
+                static_assert(std::is_base_of<T,TT>::value, "invalid type");
+                auto i = m.find(typeid(T));
+                if(i == m.end())
+                    i = m.emplace(typeid(T), container{}).first;
+				std::any_cast<container>(&i->second)->emplace_back(std::shared_ptr<T>{
+					new TT{ x },
+						[](T* p) { delete static_cast<TT*>(p); }
+				});
+            }
+        };
+
+        template <typename T> struct UniqueAttribute
+        {
+            using container = T;
+            static void add(AttrMap& m, T x)
+            {
+				if (!m.try_emplace(typeid(T), x).second) {}
+                    //throw std::runtime_error("inserting second UniqueAttribute");
+            }
+        };
+
+        template <typename T> struct AttrType {
+            using type = SimpleAttribute<T>;
+        };
+
+		struct classname { const char* name; };
+        template <> struct AttrType<classname> {
+            using type = UniqueAttribute<classname>;
+        };
+
+        struct TypeInfo;
+
+        struct TypeManager {
+            static TypeManager& get()
+            {
+                static TypeManager instance{};
+                return instance;
+            }
+            template <typename T> const TypeInfo& registerOrGetType();
+            std::unordered_map<std::string, std::shared_ptr<TypeInfo>> byName;
+            std::unordered_map<std::type_index, std::shared_ptr<TypeInfo>> byId;
```

Comment:
These maps could have better names. I suggest `nameToTypeInfo` and `idToTypeInfo`. This makes them more explicit.